google/cloud/spanner_dbapi/cursor.py

```diff
@@ -207,9 +212,31 @@ def fetchone(self):
         self._raise_if_closed()
 
         try:
-            return next(self)
+            res = next(self)
+            self._checksum.consume_result(res)
+            return res
         except StopIteration:
-            return None
+            return
+        except Aborted:
+            self.connection.retry_transaction()
+            return self.fetchone()
```

Comment:
I think this will be a problem. Assuming that this is using the `ExecuteStreamingSql` RPC, then each `next()` call could potentially mean that a new RPC is executed. So for the sake of simplicity, assume in the example below that __each__ call to `next()` executes the `ExecuteStreamingSql` RPC.

Assume the following situation:
1. The table Singers contains the following singers (last names): Allison, Morrison, Pieterson
1. The application executes the query `SELECT LastName FROM Singers ORDER BY LastName` in transaction 1.
1. The client application calls `fetchone()` which returns 'Allison'.
1. Some other transaction executes `DELETE FROM Singers WHERE LastName='Pieterson'.
1. The first transaction is aborted by the backend. A retry is executed and the retry logic checks that the checksum of the retried result set is equal to the original attempt, which it is as the first record is still 'Allison'.
1. The client application calls `fetchone()` again. This should return 'Morrison', but as it needs to call `ExecuteStreamingSql` it will (probably) use the transaction id of the original transaction (unless that transaction id has somehow been replaced in the underlying iterator). If it does use the old transaction id, the RPC will fail with yet another `Aborted` error, and that will repeat itself until the transaction retry limit has been reached.

Reply:
Seems to me we can just drop the `_transaction` property, so that `Connection` will initiate a new one on the next `execute()` call.

Reply:
@IlyaFaer @c24t 

Sorry for reopening this, and this comment should not be considered blocking for merging this PR, but I think we need to look into this once more. Only dropping the `_transaction` property will in this case not be enough for the following reason:
1. When `executeSql` is called, a streaming iterator is returned to the application.
1. That streaming iterator is linked with the transaction that was active at that moment, and a reference to that transaction is also [held in the iterator](https://github.com/googleapis/python-spanner/blob/783fffb6683119db0dde89bff66112e451881c19/google/cloud/spanner_v1/snapshot.py#L257).
1. If a transaction is aborted and the client application has consumed only parts of a streaming iterator, that iterator is no longer valid (at least: it will also throw an exception if it needs to receive more data from the server).

The JDBC driver client solves the above problem by [wrapping all streaming iterators](https://github.com/googleapis/java-spanner/blob/3fa7910c8e5089cff1c9ed645f160a9e0ddfc351/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReplaceableForwardingResultSet.java#L44) before returning these to the client application. That makes it possible for the JDBC driver to [replace the underlying streaming iterator](https://github.com/googleapis/java-spanner/blob/3fa7910c8e5089cff1c9ed645f160a9e0ddfc351/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ChecksumResultSet.java#L177) with a new one when a transaction has been aborted and successfully retried.

We should add that to the Python DBApi as well, but we could do that in a separate PR to prevent this PR from becoming even bigger than it already is.

Reply:
@c24t, @olavloite, hm-m. I think we're protected from errors here, because our connection API doesn't actually give streaming result objects to a user.

Here is where we're getting a streaming iterator:
https://github.com/googleapis/python-spanner-django/blob/196c44949370fb818e610b21c3b00344fdc3d03a/google/cloud/spanner_dbapi/cursor.py#L167-L170

So, iterator is held in the protected property `_itr`, and users will be streaming it with `Cursor.fetch*()` methods, without actual access to the iterator itself:
https://github.com/googleapis/python-spanner-django/blob/196c44949370fb818e610b21c3b00344fdc3d03a/google/cloud/spanner_dbapi/cursor.py#L204-L212

Where `next(self)` is calling `next(self._itr)` here:
https://github.com/googleapis/python-spanner-django/blob/196c44949370fb818e610b21c3b00344fdc3d03a/google/cloud/spanner_dbapi/cursor.py#L293-L296

Thus, if a transaction failed, the connection will drop the transaction, checkout a new one, re-run all the statements, each of which will replace `_itr` with new streamed iterator. So, all the iterators are processed internally, and will be replaced on a retry, as I see.